helper/include/odp/helper/threads.h

```diff
@@ -86,13 +94,127 @@ typedef struct {
+	 * thread parameters (odph_thread_param_t).
+	 */
+	odp_instance_t instance;
+
+	/**
+	 * CPU mask for thread pinning
+	 */
+	const odp_cpumask_t *cpumask;
+
+	/**
+	 * Select between Linux pthreads and processes
+	 *
+	 * 0: Use pthreads
+	 * 1: Use processes
+	 *
+	 * Default value is 0.
+	 */
+	int thread_model;
+
+	/**
+	 * Synchronized thread creation
+	 *
+	 * 0: Don't synchronize thread creation
+	 * 1: Create threads in series so that the next thread is created
+	 *    only after the previous thread have signaled that it has passed
+	 *    ODP local initialization.
+	 *
+	 * Default value is 0.
+	 */
+	int sync;
+
+	/**
+	 * Thread parameter sharing
+	 *
+	 * 0: Thread parameters are not shared. The thread parameter table
+	 *    contains 'num' elements.
+	 * 1: The thread parameter table contains a single element, which is
+	 *    used for creating all 'num' threads.
+	 *
+	 * Default value is 0.
+	 */
+	int share_param;
+
+} odph_thread_common_param_t;
+
+/**
+ * Create and pin threads (as Linux pthreads or processes)
+ *
+ * This is an updated version of odph_odpthreads_create() call. It may be called
+ * multiple times to create threads in steps. Each call launches 'num' threads
+ * and pins those to separate CPUs based on the cpumask. Use 'thread_model'
+ * parameter to select if Linux pthreads or processes are used. This selection
+ * may be overridden with ODP helper options. See e.g. --odph_proc under
+ * odph_options() documentation.
+ *
+ * Thread creation may be synchronized by setting 'sync' parameter. It
+ * serializes thread start up (odp_init_local() calls), which helps to
+ * stabilize application start up sequence.
+ *
+ * By default, the thread parameter table contains 'num' elements, one for
+ * each thread to be created. However, all threads may be created
+ * with the same parameters by setting 'share_param' parameter.
+ *
+ * Thread table must be large enough to hold 'num' elements. Also the cpumask
+ * must contain 'num' CPUs. Threads are pinned to CPUs in order - the first
+ * thread goes to the smallest CPU number of the mask, etc.
+ *
+ * Launched threads may be waited for exit with odph_thread_join(), or with
+ * direct Linux system calls.
+ *
+ * @param[out] thread        Thread table for output
+ * @param      param         Common parameters for all threads to be created
+ * @param      thr_param     Table of thread parameters
+ * @param      num           Number of threads to create
+ *
+ * @return Number of threads created
+ * @retval -1  On failure
+ *
+ * @see odph_thread_join()
+ */
+int odph_thread_create(odph_thread_t thread[],
+		       const odph_thread_common_param_t *param,
+		       const odph_thread_param_t thr_param[],
+		       int num);
+
+/**
+ * Wait previously launched threads to exit
+ *
+ * This is an updated version of odph_odpthreads_join() call. It waits for
+ * threads launched with odph_thread_create() to exit. It may be called
+ * multiple times to wait thread exits in steps.
```

Comment:
Is it allowed to wait on the same thread multiple times (as pthread allows)? I'm fine with both answers. :)

Reply:
Fixed in v2. A thread can be waited only once.